Add WaitMap structure

jamesmunns opened the pull request:
Opening this for an early architectural review. Still needs docs and (most) tests updated.

In particular:

I think some more logic regarding "wake_all" needs to be removed: DataWaitQueue doesn't support it, and I think there are some "impossible" control flow items where normally a wake_all could insert an "extra" wake into the queue.

Also, I wonder if I should allow for "failed wakes" - or wakes where NO data is available. This would change the future type to returning `Option<V>`, and would probably be needed to ACTUALLY support `close()` (which now that I think about it, probably panics if you call it now, I haven't re-enabled that test yet)

I also wonder if "WaitQueue" should/could just be a special case of `DataWaitQueue` where the data is `()`, sort of like how I've modified `WaitResult`. That being said, if I do switch to "failed wakes", I don't know if its meaningful to distinguish between `Some(())` and `None`, since a wake is a wake?

Open to thoughts! I need to go off and do other things, but interested in your early feedback on this.

jamesmunns commented:
Oh also since I wasn't super clear on it: Switching to `Option<V>` would allow a "wake_all" again, by putting `None` in all the completed futures. That being said - I'm not sure if this is useful outside something like the `close` case. The only reason I can think of would be some kind of "hey we're closing, you're never getting your data, sorry".

jamesmunns commented on maitake/src/wait/dataqueue/tests.rs:
```diff
@@ -0,0 +1,272 @@
+    //     }
+
+    #[test]
+    fn wake_one() {
```
@hawkw I did re-enable this test, which also shows sort of how you'd use this in the "normal" case. This might be worth taking a peek at first.

hawkw commented:
Some API design thoughts: when I was working on the initial `WaitQueue` implementation, my thought was that we would have something like this:
```rust
pub struct WaitQueue<T = ()> {
    // ...
}

impl<T> WaitQueue<T> {
    pub fn wake_first_match(&self, predicate: impl FnMut(&T) -> bool) {
          // ...
    }

    pub fn wake_all_matches(&self, predicate: impl FnMut(&T) -> bool) {
          // ...
    }

    pub fn wait_with(&self, data: T) -> Wait<'_, T> {
          // ...
    }

}
```
where `T` is arbitrary user-provided data that may be added to a waiter. But, at the time, I decided to punt on that for the initial implementation and add it in a follow-up.

I'm not sure if we actually _want_ the  same user-facing structure to provide both the data-less and data-full wait queue interfaces, but I thought this was worth mentioning.

Another thought, which would take this in a very different direction, is that maybe what we would want is to make this type a `WaitMap` or something, and do away with the queue-like interface entirely. That way, we could replace the linked list with an intrusive red-black tree or something later on.

jamesmunns commented:
Yeah, I think `WaitMap<K, V>` is a much better name for what I have in mind. In particular, I think the interface already matches that (it only provides a `wake(&K)` interface).

It's not really enforced atm, but the idea is that all `K` instances would be unique. We could handle that one of a few ways:

* Actually check, meaning that `wait()` becomes `O(n)` instead of O(1) - (though it probably stays log(n) if we use a tree), and becomes failable
* Don't check, and state that if you have duplicate `K`s in the map, one of them will be awoken and give the data. we could probably say "first inserted" if it's a queue, it would be harder if the underlying structure is a map.

I'm inclined to do the former and PRETEND we have a tree under the hood (when really it's a list), meaning:

* Today:
    * Insertion is O(1)
    * Wake is O(n)
* With a tree later:
    * Insertion is O(log(n))
    * Wake is O(log(n))

jamesmunns commented:
Reiterating from chat:

I plan to make the following changes/executive decisions:

* Change the name to WaitMap
* Make `wait(K)` check if the key exists, and panic if it does
* Add a `try_wait(K)` that checks and returns an err if it exists
* Change the wait future to resolve to an `Option<V>`
* Make sure all the data paths still make sense in the absence of a `wake_all` function (though still work with `close()`).

Lemme know if you think of any reason these choices wouldn't be acceptable!

hawkw commented on maitake/src/wait/dataqueue.rs:
```diff
@@ -0,0 +1,984 @@
+/// The queue's current state.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+#[repr(u8)]
+enum State {
+    /// No waiters are queued, and there is no pending notification.
+    /// Waiting while the queue is in this state will enqueue the waiter;
+    /// notifying while in this state will store a pending notification in the
+    /// queue, transitioning to [`State::Woken`].
+    Empty = 0b00,
+
+    /// There are one or more waiters in the queue. Waiting while
+    /// the queue is in this state will not transition the state. Waking while
+    /// in this state will wake the first waiter in the queue; if this empties
+    /// the queue, then the queue will transition to [`State::Empty`].
+    Waiting = 0b01,
+
+    /// The queue has a stored notification. Waiting while the queue
+    /// is in this state will consume the pending notification *without*
+    /// enqueueing the waiter and transition the queue to [`State::Empty`].
+    /// Waking while in this state will leave the queue in this state.
+    Woken = 0b10,
+
+    /// The queue is closed. Waiting while in this state will return
+    /// [`Closed`] without transitioning the queue's state.
+    ///
+    /// *Note*: This *must* correspond to all state bits being set, as it's set
+    /// via a [`fetch_or`].
+    ///
+    /// [`Closed`]: crate::wait::Closed
+    /// [`fetch_or`]: core::sync::atomic::AtomicUsize::fetch_or
+    Closed = 0b11,
+}
+
+#[derive(Clone, Debug)]
+enum Wakeup {
+    Empty,
+    Waiting(Waker),
+    One,
+    Closed,
+}
```
it seems like some of these internals could potentially be shared with `WaitQueue`?

hawkw commented on maitake/src/wait/dataqueue.rs:
```diff
@@ -0,0 +1,984 @@
+    }
+}
+
+bitfield! {
+    #[derive(Eq, PartialEq)]
+    struct QueueState<usize> {
+        /// The queue's state.
+        const STATE: State;
+    }
+}
+
+bitfield! {
+    #[derive(Eq, PartialEq)]
+    struct WaitStateBits<usize> {
+        /// The waiter's state.
+        const STATE: WaitState;
+    }
+}
```
since these are _just_ enums, and don't also contain a count of `wake_all` calls, it seems like we don't need the bitfield and can just convert the `AtomicUsize` directly into a `State` (for the queue), or just store the enum in the waiter directly (for the waiter).

jamesmunns replied:
Yeah, makes sense! I'll address this in the next chunk of work.

hawkw commented on maitake/src/wait/dataqueue.rs:
```diff
@@ -0,0 +1,984 @@
+        loop {
+            match state.get(QueueState::STATE) {
+                // if the queue is closed, bail.
+                State::Closed => return WakeOutcome::NoMatch(val),
```
is this really `NoMatch`? perhaps there should be an additional `WakeOutcome` variant for "queue is already closed"?

jamesmunns replied:
No objections! "NoMatch" really means "I didn't wake anyone, here's your data back", but might be worth distinguishing.

hawkw commented on maitake/src/wait/dataqueue.rs:
```diff
@@ -0,0 +1,984 @@
+    }
+}
+
+#[repr(C)]
+struct Node<K: PartialEq, V> {
+    /// Intrusive linked list pointers.
+    ///
+    /// # Safety
+    ///
+    /// This *must* be the first field in the struct in order for the `Linked`
+    /// impl to be sound.
+    links: list::Links<Waiter<K, V>>,
+
+    /// The node's waker
+    waker: Wakeup,
+
+    // This type is !Unpin due to the heuristic from:
+    // <https://github.com/rust-lang/rust/pull/82834>
+    _pin: PhantomPinned,
+}
```
this is identical to the `Node` in `WaitQueue`...it seems like we could just make this generic over the `Waiter` type, and reuse this code (so the waitmap would have a `Node<Waiter<K, V>>` while `WaitQueue` would have a `Node<Waiter>`).

jamesmunns replied:
Seems reasonable!

jamesmunns commented on maitake/src/wait/dataqueue.rs:
```diff
@@ -0,0 +1,984 @@
+    queue: Mutex<List<Waiter<K, V>>>,
+
+    _key: PhantomData<K>,
+    _val: PhantomData<V>,
```
Correct! Will remove now (and the custom debug impl).

jamesmunns commented on maitake/src/wait/dataqueue.rs:
```diff
@@ -0,0 +1,984 @@
+        // snapshot the queue's current state.
+        let mut state = self.load();
+
+        // check if any tasks are currently waiting on this queue. if there are
+        // no waiting tasks, store the wakeup to be consumed by the next call to
+        // `wait`.
+        loop {
+            match state.get(QueueState::STATE) {
+                // if the queue is closed, bail.
+                State::Closed => return WakeOutcome::NoMatch(val),
+                // if there are waiting tasks, break out of the loop and wake one.
+                State::Waiting => break,
+                _ => {}
+            }
+
+            let next = state.with_state(State::Woken);
+            // advance the state to `Woken`, and return (if we did so
+            // successfully)
+            match self.compare_exchange(state, next) {
+                Ok(_) => return WakeOutcome::NoMatch(val),
+                Err(actual) => state = actual,
+            }
+        }
```
Yeah, this is the "data paths that don't make sense". Agreed these should be cleared up (code so far has been pretty "hack and slash").

hawkw commented on maitake/src/wait/dataqueue.rs:
```diff
@@ -166,6 +166,7 @@ mod tests;
 /// [ilist]: cordyceps::List
 /// [intrusive]: https://fuchsia.dev/fuchsia-src/development/languages/c-cpp/fbl_containers_guide/introduction
 /// [2]: https://www.1024cores.net/home/lock-free-algorithms/queues/intrusive-mpsc-node-based-queue
+#[derive(Debug)]
```
nit: this debug impl still requires `K` and `V` to be `Debug`, which it maybe shouldn't?

jamesmunns replied:
Does it? Waiter has a custom debug impls, which I think covers it? But I'll make sure before marking this ready.

hawkw commented on maitake/src/wait/map.rs:
```diff
@@ -458,25 +442,38 @@ impl<K: PartialEq, V> DataWaitQueue<K, V> {
+        let mut mg = self.queue.lock();
+        let mut cursor = mg.cursor();
+        if cursor.find(|n| n.key == key).is_some() {
+            return Err(());
```
do you think this should return the key value?

jamesmunns replied:
Maybe! Probably couldn't hurt.

jamesmunns commented:
So, a little more progress.

I think I've cleaned up most of the vestigial wake-all state, which should be good.

I also changed my mind again, the Waiter SHOULD resolve to `Result<V>`, not `Result<Option<V>>`, because the only way to wake without data is by closing, and that's an error already.

I *also* am pretty sure I need to add a drop impl somewhere - if we HAVE a V, and the future drops, I'm pretty sure we just drop it without running a dtor, because it's in a MaybeUninit. I'll add a test or three for this.

jamesmunns commented:
Okay @hawkw, I'd like to call this one ready to merge now.

I believe I have addressed all comments, EXCEPT for refactoring `WaitMap` and `WaitQueue` to have some common internals. If you don't have any argument, I'd like to defer that to a later PR, as I'm a little out of brain energy for this at the moment.

Happy to address any issues or "must fix" review comments!

hawkw commented:
> I believe I have addressed all comments, EXCEPT for refactoring `WaitMap` and `WaitQueue` to have some common internals. If you don't have any argument, I'd like to defer that to a later PR, as I'm a little out of brain energy for this at the moment.

I'm totally fine with saving that for later!

hawkw commented on maitake/src/wait/map.rs:
```diff
@@ -0,0 +1,926 @@
+}
+
+/// An indication of the result of a call to `wait()`
+// Note: We duplicate the WaitResult type as it currently only has "Closed"
+// as an error condition. We should merge these two on the next breaking
+// change.
```
since this crate has never been published, it's fine to do that now --- although, honestly, i think it's fine (and perhaps better) for the `WaitMap` type to have a different error type than the other types in the wait module. two of those error conditions can never occur with `WaitQueue` or `WaitCell`, so i have a slight preference for them returning an error type where users will not have to match the error and handle two cases that will never happen.

jamesmunns replied:
Makes sense! Will remove the comment.

hawkw commented on maitake/src/wait/map.rs:
```diff
@@ -0,0 +1,926 @@
+mod tests;
+
+/// An error indicating a failed wake
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
```
this probably should have a 
```suggestion
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
#[non_exhaustive]
```
on it, if there's a possibility of adding new error types? not a blocker.

hawkw commented on maitake/src/wait/map.rs:
````diff
@@ -0,0 +1,926 @@
+///
+/// // Spawn some tasks that will wait on the queue.
+/// // We'll use the task index (0..10) as the key.
+/// for i in 0..TASKS {
+///     let q = q.clone();
+///     scheduler.spawn(async move {
+///         let val = q.wait(i).await.unwrap();
+///         assert_eq!(val, i + 100);
+///     });
+/// }
+///
+/// // Tick the scheduler once.
+/// let tick = scheduler.tick();
+///
+/// // No tasks should complete on this tick, as they are all waiting
+/// // to be woken by the queue.
+/// assert_eq!(tick.completed, 0, "no tasks have been woken");
+///
+/// // We now wake each of the tasks, using the same key (0..10),
+/// // and provide them with a value that is their `key + 100`,
+/// // e.g. 100..110. Only the task that has been woken will be
+/// // notified.
+/// for i in 0..TASKS {
+///     let result = q.wake(&i, i + 100);
+///     assert!(matches!(result, WakeOutcome::Woke));
+///
+///     // Tick the scheduler.
+///     let tick = scheduler.tick();
+///
+///     // Exactly one task should have completed
+///     assert_eq!(tick.completed, 1);
+/// }
+///
+/// // Tick the scheduler.
+/// let tick = scheduler.tick();
+///
+/// // No additional tasks should be completed
+/// assert_eq!(tick.completed, 0);
+/// assert!(!tick.has_remaining);
+/// ```
````
nitpicky, not a big deal: is there a similar test in the regular `tests` module? i think doctests won't get run with miri etc.

jamesmunns replied:
Yup! This is basically a duplicate of `maitake::wait::map::tests::wake_one()`.

hawkw commented on maitake/src/wait/map.rs:
```diff
@@ -0,0 +1,926 @@
+#[pinned_drop]
+impl<K: PartialEq, V> PinnedDrop for Waiter<K, V> {
+    fn drop(self: Pin<&mut Self>) {
+        // SAFETY: If we are being dropped, then we have already been removed
+        // from the map via Waiter::release(). This means we now have exclusive
+        // access to the node, and no lock is required.
```
(or we were never in the map to begin with)

hawkw commented on maitake/src/wait/map.rs:
```diff
@@ -0,0 +1,926 @@
+    /// corresponds to `WaitState::Waiting`.
+    Waiting(Waker),
+
+    /// The Waiter has received data, and is waiting for the woken task
+    /// to notice, and take the data by polling+completing the future.
+    /// This corresponds to `WaitState::Completed`.
+    DataReceived,
```
a thought (which is fine to save for a follow-up): i feel like it would make sense to actually store the data in this variant, instead of having a separate `UnsafeCell<MaybeUninit<V>>`? since the `Wakeup` is essentially responsible for determining whether there's a value in there or not. i don't _think_ there's a performance/space improvement from not just storing it here?

jamesmunns replied:
That's a good point! I've added a TODO comment.

hawkw commented on maitake/src/wait/map.rs:
```diff
@@ -0,0 +1,926 @@
+        if let Some(mut node) = self.node_match_locked(key, &mut *queue, state) {
+            let waker = Waiter::<K, V>::wake(node, &mut *queue, Wakeup::DataReceived);
+            unsafe {
+                node.as_mut().val.with_mut(|v| (*v).write(val));
```
this is safe because we're holding the lock...maybe worth a comment?

hawkw replied:
(if we stored the data in the `Node`, we could use `Waiter::with_node` to "prove" we're holding the lock...but maybe that's not worth it)

hawkw commented on maitake/src/wait/map.rs:
```diff
@@ -0,0 +1,926 @@
+    /// woken by a call to [`wake`] with a matching `key`, or when the `WaitMap`
+    /// is dropped.
+    ///
+    /// NOTE: `key`s must be unique. If the given key already exists in the
```
very unimportant nit: usually i bold these
```suggestion
    /// **Note**: `key`s must be unique. If the given key already exists in the
```

hawkw commented on maitake/src/wait/map.rs:
```diff
@@ -0,0 +1,926 @@
+    /// woken by a call to [`wake`] with a matching `key`, or when the `WaitMap`
+    /// is dropped.
+    ///
+    /// NOTE: `key`s must be unique. If the given key already exists in the
```
```suggestion
    /// **Note**: `key`s must be unique. If the given key already exists in the
```
(but i don't actually care)

hawkw commented on maitake/src/wait/map.rs:
```diff
@@ -0,0 +1,926 @@
+    /// [`wake`]: Self::wake
+    pub fn try_wait(&self, key: K) -> Result<Wait<'_, K, V>, K> {
+        // Check if key already exists
+        let mut mg = self.queue.lock();
```
unimportant style nit: i found the name `mg` kind of un-obvious, but :woman_shrugging:

jamesmunns replied:
Reasonable! Switched to `guard`.

hawkw commented on maitake/src/wait/map.rs:
```diff
@@ -0,0 +1,926 @@
+    }
+}
+
+pub enum WakeOutcome<V> {
+    Woke,
+    NoMatch(V),
+    Closed(V),
+}
```
these should probably have some quick comments?

hawkw commented on maitake/src/wait/map.rs:
```diff
@@ -0,0 +1,926 @@
+
+                            notified(val)
+                        }
+                        Wakeup::Retreived => {
+                            // TODO: Should I change the result type to also return an error
+                            // for this case? Right now error can only be `Closed`
+                            consumed()
+                        }
```
it looks like the TODO comment here was addressed?
```suggestion
                        Wakeup::Retreived => consumed(),
```

hawkw commented on maitake/src/wait/map/tests.rs:
```diff
@@ -0,0 +1,368 @@
+        for i in 0..TASKS {
+            let q = q.clone();
+            scheduler.spawn(async move {
+                // TODO(AJM): I need to select!() against a one shot channel or something
+                // for the "wait hanging" test
```
did you want to address this before merging?

jamesmunns replied:
Good catch! I already added the `drop_wake_bailed()` test, which addresses this.

hawkw commented on maitake/src/wait/map/tests.rs:
```diff
@@ -0,0 +1,368 @@
+        let tick = scheduler.tick();
+
+        assert_eq!(tick.completed, 0);
+        assert_eq!(COMPLETED.load(Ordering::SeqCst), 0);
+        assert_eq!(KEY_DROPS.load(Ordering::SeqCst), 0);
+        assert_eq!(VAL_DROPS.load(Ordering::SeqCst), 0);
+        assert!(!tick.has_remaining);
+
+        q.close();
+
+        let tick = scheduler.tick();
+
+        assert_eq!(tick.completed, TASKS);
+        assert_eq!(COMPLETED.load(Ordering::SeqCst), TASKS);
+        assert_eq!(KEY_DROPS.load(Ordering::SeqCst), TASKS);
+        assert_eq!(VAL_DROPS.load(Ordering::SeqCst), 0);
```
nit, take it or leave it: all the atomic ops in these tests *can* be `Relaxed` --- the tests are all single-threaded, so these atomics will never be modified anywhere else, so ordering doesn't matter at all. also, they're just counters, so they can always be `Relaxed`. not a big deal though.

hawkw reviewed (approved):
looks good to me overall! i left a few minor suggestions.

jamesmunns commented:
Addressed all comments, marking this for auto-merge if CI passed.
